### PR TITLE
[KEYCLOAK-6116] - Get email attribute from 'subject alternative name' using X509 certificate

### PR DESCRIPTION
**Pull request by pedroigor**
Still need to work on tests. Need to figure out someway to include this scenario in X509 testsuite. It seems tests are really dependent on a single client certificate with no option to support different certificates.

### CODE REVIEW

**Review by stianst — DISMISSED** [review]
Needs tests to be added before merging

**Comment by pedroigor** [comment]
@stianst , I've added tests.

In order to be able to test this scenario and provide a better configuration for X509 related tests, I've also added a OpenSSL Certificate Authority. We should be able now to create certs more easily, using any extension we may need, check for certificate chains more easily, etc.

**Review by mposolda — DISMISSED** [review]

**Comment by mposolda** [comment]
@pedroigor I am adding approve. Sorry it took so long. Will try to be better next time :)

I am adding @vmuzikar as another reviewer. AFAIK he is responsible for X509 testing on QE side. Even if he doesn't have concerns, it may be good that at least he is aware of those changes :) Especially PhantomJS setup changes and maybe other things too.

**Comment by vmuzikar** [comment]
@pedroigor Will look into it first thing tomorrow. Would that be ok?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509/AbstractX509AuthenticationTest.java [hunk 1]
```diff
@@ -100,6 +102,27 @@ protected boolean isImportAfterEachMethod() {
         return true;
     }
 
+    @BeforeClass
+    public static void onBeforeTestClass() {
+        if (System.getProperty("auth.server.container", "undefined").endsWith("wildfly")) {
```
Comment (vmuzikar) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509/AbstractX509AuthenticationTest.java]: This won't work wit auth-server-eap.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509/X509BrowserLoginSubjectAltNameEmailTest.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Page
+    protected X509IdentityConfirmationPage loginConfirmationPage;
+
+    @Page
+    protected LoginPage loginPage;
+
+    @BeforeClass
+    public static void onBeforeTestClass() {
+        if (System.getProperty("auth.server.container", "undefined").endsWith("wildfly")) {
```
Comment (vmuzikar) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509/X509BrowserLoginSubjectAltNameEmailTest.java]: This won't work wit auth-server-eap.

**Review by vmuzikar — CHANGES_REQUESTED** [review]
The X.509 tests won't work with EAP, i.e. RH-SSO. Could you please fix that?

Just my opinion: The new phantom.js config looks to me a bit unnecessary complicated - lots of copy-paste and commands just to change cert, key and passphrase. The main problem I see here is that we need to rewrite all the phantom.js configs just to use a client cert. Imagine we would need to change some global phantom.js config - then we'd need to think of all the tests which are using certs, and make the changes there as well. Couldn't we make some dedicated system props just for the cert, key and pass?
What do you think?

**Comment by vmuzikar** [comment]
Also, should we update docs with this as well?

**Comment by pedroigor** [comment]
@vmuzikar, here it is https://github.com/keycloak/keycloak-documentation/pull/334.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509/AbstractX509AuthenticationTest.java (hunk 1 above)
Comment (pedroigor) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509/AbstractX509AuthenticationTest.java]: True, forgot about eap ... I guess it should be enough to include a check of "eap" there ?

Comment (vmuzikar) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509/AbstractX509AuthenticationTest.java]: I believe something like that would work too:
```java
Boolean.parseBoolean(System.getProperty("auth.server.jboss"));
```

**Comment by pedroigor** [comment]
@vmuzikar, X509 tests are extending from ```AbstractX509AuthenticationTest```. 

The changes related with phantomjs is a workaround to overcome phantomjs config and execution in testsuite.

I would say that you could merge this PR and have a separated issue to think about how to properly handle these issues.

**Review by vmuzikar — APPROVED** [review]
@pedroigor Ok, LGTM then. Created https://issues.jboss.org/browse/KEYCLOAK-6792

**Comment by pedroigor** [comment]
@stianst, can we have your bless on this one ?

**Review by stianst — APPROVED** [review]

**Pull request merged**